contracts/axelar-auth-verifier/src/test.rs

```diff
@@ -44,6 +106,116 @@ fn test_validate_proof() {
     assert!(latest_signer_set);
 }
 
+#[test]
+#[should_panic(expected = "invalid epoch")]
+fn test_fail_validate_proof_invalid_epoch() {
+    let (env, _, client) = setup_env();
+    let user = Address::generate(&env);
+
+    initialize(&env, &client, user, randint(0, 10), randint(1, 10));
+
+    let different_signers = generate_signer_set(&env, randint(1, 10), false);
+
+    let msg = Bytes::from_array(&env, &[0x01, 0x02, 0x03]);
+    let msg_hash = env.crypto().keccak256(&msg);
+    let proof = generate_proof(&env, msg_hash.clone(), different_signers);
+
+    // should panic, epoch should return zero for invalid signer set
+    client.validate_proof(&msg_hash, &proof.to_xdr(&env));
+}
+
+#[test]
+#[should_panic(expected = "invalid signatures")]
+fn test_fail_validate_proof_invalid_signatures() {
+    let (env, _, client) = setup_env();
+    let user = Address::generate(&env);
+
+    let signers = initialize(&env, &client, user, randint(0, 10), randint(1, 10));
+
+    let msg = Bytes::from_array(&env, &[0x01, 0x02, 0x03]);
+    let msg_hash = env.crypto().keccak256(&msg);
+    let proof = generate_proof(&env, msg_hash.clone(), signers);
+
+    let different_msg = Bytes::from_array(&env, &[0x04, 0x05, 0x06]);
+    let different_msg_hash = env.crypto().keccak256(&different_msg);
+
+    // should panic, proof is for different message hash
+    client.validate_proof(&different_msg_hash, &proof.to_xdr(&env));
+}
+
+#[test]
+#[should_panic(expected = "invalid signatures")]
+fn test_fail_validate_proof_empty_signatures() {
+    let (env, _, client) = setup_env();
+    let user = Address::generate(&env);
+
+    let signers = initialize(&env, &client, user, randint(0, 10), randint(1, 10));
+
+    let msg = Bytes::from_array(&env, &[0x01, 0x02, 0x03]);
+    let msg_hash = env.crypto().keccak256(&msg);
+    let mut proof = generate_proof(&env, msg_hash.clone(), signers);
+
+    proof.signatures = Vec::new(&env);
+
+    // validate_proof should panic, empty signatures
+    client.validate_proof(&msg_hash, &proof.to_xdr(&env));
+}
+
+#[test]
+#[should_panic(expected = "invalid signatures")]
+fn test_fail_validate_proof_invalid_signer_set() {
+    let (env, _, client) = setup_env();
+    let user = Address::generate(&env);
+
+    let signers = initialize(&env, &client, user, randint(0, 10), randint(1, 10));
+
+    let msg = Bytes::from_array(&env, &[0x01, 0x02, 0x03]);
+    let msg_hash = env.crypto().keccak256(&msg);
```

Comment:
repetition, create a util to generate random payload and hash. you can use `BytesN::<10>::random(&env);`